[RFR] Integration tests for Redfish Physical Chassis and Racks

matejart opened the pull request:
This PR adds support for Physical Chassis and Physical Racks, which are parts of the Redfish provider's Physical Infrastructure System.

In this PR, we
* Fix the use of `PhsyicalRackEntity`.
* Extend tests in the Redfish Physical Infrastructure System to validate the number of Physical Chassis and Physical Racks in the Details page.
* Extend physical rack functionality to enable tests for inventory of Redfish phsyical racks.
* Add the implementation of the elements in the GUI representing and showing physical infrastructures' physical chassis.
* Add testing of the Redfish Physical Chassis GUI. We test for chassis name, description and current status of the chassis identifying LED, and we check the count of physical systems enclosed within the chassis.
    
Related Bugzilla tickets:
   * https://bugzilla.redhat.com/show_bug.cgi?id=1571610
   * https://bugzilla.redhat.com/show_bug.cgi?id=1574809

Depends on PRs:
  * https://github.com/ManageIQ/wrapanapi/pull/335 (merged)
  * https://github.com/ManageIQ/integration_tests/pull/7726 (merged)

matejart commented:
This PR is now ready for internal tests. @miha-plesko can you please provide your feedback?

miha-plesko commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,294 @@
+    title = Text('.//div[@id="center_div" or @id="main-content"]//h1')
+
+    @property
+    def in_compute_physical_infrastructure_chassis(self):
```
Isn't this usually called `is_displayed`?

miha-plesko replied:
EDIT: I see, so you want to evaluate it from subclasses. But still, I think we should rename this method into

```
def is_displayed(self):
  ...
```

like in any other View and call it from subclasses with

```
super(self.__class__, self).is_displayed
```

@mshriver would you agree, or should we go with custom "is_displayed" function names?

miha-plesko commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,294 @@
+    @property
+    def is_displayed(self):
+        title = "{name} (Summary)".format(name=self.context["object"].name)
+        return (self.in_compute_physical_infrastructure_chassis and
```
As mentioned in the comment above, we could refer to `is_displayed` of the parent via `super(self.__class__, self).is_displayed` instead implementing custom function names.

miha-plesko commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,294 @@
+    @property
+    def data(self):
+        data_dict = super(PhysicalChassisEntity, self).data
+        if 'quadicon' in data_dict and data_dict['quadicon']:
```
```suggestion
        if data_dict.get('quadicon', ''):
```

miha-plesko replied:
Also, please decide either for single quotes or double. You're mixing them from function to function :trollface:

matejart replied:
👍

miha-plesko commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,294 @@
+
+class PhysicalChassisDetailsEntities(View):
+    """Represents Details page Entities."""
+    properties = SummaryTable(title="Properties")
```
SummaryTable approach is deprecated in favor of ParametrizedSummaryTable, see here for example https://github.com/ManageIQ/integration_tests/pull/6802. In general, this means you needn't list every piece separately but rather only:

```
summary = ParametrizedSummaryTable
```

and then use it like https://github.com/ManageIQ/integration_tests/blob/4757297e7e3bcc90b747470636b7d958e360857f/cfme/networks/provider/nuage.py#L124-L127

matejart replied:
👍

miha-plesko commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,294 @@
+    @variable(alias='ui')
+    def chassis_name(self):
+        view = navigate_to(self, "Details")
+        return view.entities.properties.get_text_of("Chassis name")
```
Please see my other comment on this topic, but I think approach with multiple entities is deprechated. Try go with:

```suggestion
        return view.entities.summary("Properties").get_text_of("Chassis name")
```

matejart replied:
👍

miha-plesko commented on cfme/tests/physical_infrastructure/test_redfish_physical_chassis.py:
```diff
@@ -0,0 +1,20 @@
+
+
+@pytest.fixture(scope="function")
+def physical_chassis(appliance, provider, setup_provider_funcscope):
```
Out of curiosity: what does `setup_provider_funcscope` do? Sounds like we're forcibly creating provider everytime, not sure if that's intended?

matejart replied:
As [required](https://github.com/ManageIQ/integration_tests/pull/7726#discussion_r224406426), these fixtures are restricted to funcion scope, thus `funscope` suffix instead of `modscope`.

miha-plesko commented on cfme/tests/physical_infrastructure/test_redfish_physical_chassis.py:
```diff
@@ -0,0 +1,20 @@
+    yield appliance.collections.redfish_physical_chassis.all(provider)
+
+
+def test_redfish_physical_chassis_details_stats(physical_chassis):
```
So the fixture is in singluar `physical_chassis` but it returns a list of them... Maybe rename it to `physical_chasis_list` to emphasise it's a list.

matejart replied:
Actually, you can't tell if "chassis" is singular or plural from the spelling alone. That is why the docstring says it's a collection.

miha-plesko commented on cfme/tests/physical_infrastructure/test_redfish_physical_racks.py:
```diff
@@ -0,0 +1,18 @@
+@pytest.fixture(scope="function")
+def physical_racks(appliance, provider, setup_provider_funcscope):
+    """Get and return the physical racks collection."""
+    yield appliance.collections.redfish_physical_racks.all(provider)
```
Not sure if it makes sense to put this into a fixture, but okay

mshriver replied:
@miha-plesko I agree here and would actually request that this is removed and this call made during the test. There's no setup occurring here, and I've recently made the same request to other PR's placing collections within fixtures. The test or any fixture requiring the collection and its entities will have the `appliance` fixture readily available from which to generate the list.

matejart commented:
Thank you for your comments, @miha-plesko.

I have addressed most of the comments, so the PR is ready for QE review.
Please note that travis is failing because I raised the requirement of wrapanapi for https://github.com/ManageIQ/wrapanapi/pull/335.

mshriver commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,294 @@
+                    "Server inventory information does not contain '{}'".format(inventory))
+            except AttributeError:
+                msg = "Provider does not know how to get '{}'"
+                raise ProviderHasNoProperty(msg.format(inventory))
```
Yeah there is a lot of duplication here, appears to be the exact same code blocks for stats and inventory, that could be collapsed to iterate over the match types. Can be a later enhancement though, I don't think this is unreasonable per se.

mshriver commented on cfme/physical/physical_chassis.py:
```diff
@@ -0,0 +1,291 @@
+
+    name = attr.ib()
+    ems_ref = attr.ib(default=None)
+    provider = attr.ib(default=None)
```
This is going to seem like a nitpick, and I was hoping to merge this today, but `provider` really needs to come before `ems_ref` for the instance attributes to maintain consistency when instantiating the entities.

matejart replied:
👍

mshriver commented on cfme/physical/physical_rack.py:
```diff
@@ -232,7 +253,7 @@ def is_displayed(self):
 
 class PhysicalRacksToolbar(View):
-    """Represents hosts toolbar and its controls."""
+    """Represents racks toolbar and its controls."""
```
hah, found some copypasta :)

mshriver commented on requirements/frozen.py3.txt:
```diff
@@ -305,7 +305,7 @@ Werkzeug==0.14.1
 widgetastic.patternfly==0.0.39
 widgetsnbextension==3.4.2
-wrapanapi==3.0.30
+wrapanapi==3.0.32
```
3.0.33 is available and includes a recently merged refactor of some of the RedFish objects.

matejart replied:
👍
